spec/services/get_products_service_spec.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe GetProductsService do
+  subject(:service) { described_class.new(category_name) }
+
+  let(:category_name) { 'Cà phê'.downcase }
+  let(:all_category_name) { 'Tất cả'.downcase }
+
+  let!(:parent_category) { create(:category, name: :all_category_name) }
+  let!(:first_category) { create(:category, name: category_name, parent: parent_category) }
+  let!(:second_category) { create(:category, parent: parent_category) }
+
+  let!(:first_product) { create(:product, category: first_category) }
+  let!(:sizes) { create_list(:size, 1, product: first_product) }
+  let!(:toppings) { create_list(:topping, 1, product: first_product) }
+  let!(:tag) { create(:tag, product: first_product) }
+
+  let!(:second_product) { create(:product, category: second_category) }
+  let!(:another_sizes) { create_list(:size, 1, product: second_product) }
+  let!(:another_toppings) { create_list(:topping, 1, product: second_product) }
+  let!(:another_tag) { create(:tag, product: second_product) }
+
+  describe '#validate' do
+    context 'when category_name does not exist in db' do
+      let(:another_category_name) { Faker::Dessert.unique.variety }
+      let(:first_category) { create(:category, name: another_category_name) }
+
+      it 'returns error' do
+        service.call
+
+        expect(service.error?).to eq true
+        expect(service.first_error.message).to eq 'category is not found.'
+      end
+    end
+
+    describe '#call' do
+      context 'when validation succeeds single category' do
+        it 'returns products of first_category' do
+          service.call
+
+          expect(service.result.first.name).to eq first_category.name
```

Comment:
should check for more products under that 'category' this spec only show 1 case get single category by name, therefore it does not detect the query is wrong